dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/DefaultTrackerProgramRuleService.java

```diff
@@ -0,0 +1,105 @@
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+import lombok.extern.slf4j.Slf4j;
+import org.hisp.dhis.program.ProgramInstance;
+import org.hisp.dhis.program.ProgramStageInstance;
+import org.hisp.dhis.programrule.engine.DefaultProgramRuleEngineService;
+import org.hisp.dhis.rules.models.RuleEffect;
+import org.hisp.dhis.tracker.bundle.TrackerBundle;
+import org.hisp.dhis.tracker.converter.TrackerConverterService;
+import org.hisp.dhis.tracker.domain.Enrollment;
+import org.hisp.dhis.tracker.domain.Event;
+import org.hisp.dhis.tracker.preheat.TrackerPreheat;
+import org.springframework.stereotype.Service;
+
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+/**
+ * @author Enrico Colasante
+ */
+@Slf4j
+@Service
+public class DefaultTrackerProgramRuleService
+    implements TrackerProgramRuleService
+{
+    private final DefaultProgramRuleEngineService programRuleEngineService;
+
+    private final TrackerConverterService<Enrollment, ProgramInstance> enrollmentTrackerConverterService;
+
+    private final TrackerConverterService<Event, ProgramStageInstance> eventTrackerConverterService;
+
+    public DefaultTrackerProgramRuleService( DefaultProgramRuleEngineService programRuleEngineService,
+        TrackerConverterService<Enrollment, ProgramInstance> enrollmentTrackerConverterService,
+        TrackerConverterService<Event, ProgramStageInstance> eventTrackerConverterService )
+    {
+        this.programRuleEngineService = programRuleEngineService;
+        this.enrollmentTrackerConverterService = enrollmentTrackerConverterService;
+        this.eventTrackerConverterService = eventTrackerConverterService;
+    }
+
+    @Override
+    public Map<Enrollment, List<RuleEffect>> calculateEnrollmentRuleEffects( TrackerBundle trackerBundle )
+    {
+        return getProgramRulesForEnrollments( trackerBundle.getEnrollments(), trackerBundle.getPreheat() );
+    }
+
+    @Override
+    public Map<Event, List<RuleEffect>> calculateEventRuleEffects( TrackerBundle trackerBundle )
+    {
+        return getProgramRulesForEvents( trackerBundle.getEvents(), trackerBundle.getPreheat() );
+    }
+
+    private Map<Enrollment, List<RuleEffect>> getProgramRulesForEnrollments( List<Enrollment> enrollments,
+        TrackerPreheat preheat )
+    {
+        return enrollments
+            .parallelStream()
```

Comment:
Just make sure when using `parallelStream` that the objects the system is iterating over do not need to lazily access the db. Proxies need a Hibernate `Session` to access the db lazily, but Session is not thread-safe and you'll end up with very weird, random errors.

Reply:
Good catch.
Removed parallelStream, we will need to improve the code to build the context to call rule-engine for all the events/enrollments and then maybe use parallelStream to make the calls to rule-engine library